MergerCli/Process.cs

```diff
@@ -4,57 +4,86 @@
             long totalTileCount = newData.TileCount();
+            batchStatusManager.InitializeLayer(newData.Path);
             long tileProgressCount = 0;
-
+            bool resumeMode = false;
+            bool pollForBatch = true;
             string? resumeBatchIdentifier = batchStatusManager.GetLayerBatchIdentifier(newData.Path);
             if (resumeBatchIdentifier != null)
             {
-                newData.setBatchIdentifier(resumeBatchIdentifier);
+                resumeMode = true;
+                this._logger.LogDebug($"Resume mode activated");
                 // fix resume progress bug for gpkg, fs and web, fixing it for s3 requires storing additional data.
                 if (newData.Type != DataType.S3)
                 {
-                    tileProgressCount = int.Parse(resumeBatchIdentifier);
+                    long totalCompletedTiles = batchStatusManager.GetTotalCompletedTiles(newData.Path);
+                    tileProgressCount = totalCompletedTiles;
                 }
             }
 
             this._logger.LogInformation($"Total amount of tiles to merge: {totalTileCount - tileProgressCount}");
-
-            bool uploadOnly = _config.GetConfiguration<bool>("GENERAL", "uploadOnly");
+            var uploadOnly = this._configManager.GetConfiguration<bool>("GENERAL", "uploadOnly");
             _getTileByCoord = uploadOnly || baseData.IsNew ?
                 (_) => null
                 :
                 (targetCoords) => baseData.GetCorrespondingTile(targetCoords, true);
 
-            do
+            ParallelRun(targetFormat, baseData, newData, batchStatusManager,
+                tileProgressCount, totalTileCount, resumeBatchIdentifier, resumeMode, pollForBatch);
+            
+            batchStatusManager.CompleteLayer(newData.Path);
+            newData.Reset();
+            // base data wrap up is in program as the same base data object is used in multiple calls 
+        }
+
+        private void DoWork(TileFormat targetFormat, IData baseData, IData newData,
+            BatchStatusManager batchStatusManager, ref long tileProgressCount, long totalTileCount, bool resumeMode,ref bool pollForBatch, string? incompleteBatchIdentifier = null)
+        {
+            List<Tile> newTiles;
+            ConcurrentBag<Tile> tiles = new ConcurrentBag<Tile>();
+            string? currentBatchIdentifier;
+            string? nextBatchIdentifier;
+            lock (_locker)
             {
-                List<Tile> newTiles = newData.GetNextBatch(out string batchIdentifier);
-                batchStatusManager.SetCurrentBatch(newData.Path, batchIdentifier);
+                if (resumeMode && incompleteBatchIdentifier is not null)
+                {
+                    newData.setBatchIdentifier(incompleteBatchIdentifier);
+                }
+                
+                newTiles = newData.GetNextBatch(out currentBatchIdentifier, out nextBatchIdentifier, incompleteBatchIdentifier, totalTileCount);
```

Comment:
We should talk again about sending `incompleteBatchIdentifier` to `GetNextBatch`.